### PR TITLE
Implement proper signal handling in select calls.

### PR DESCRIPTION
**Pull request by DimCitus**
See #392.

### CODE REVIEW

**Review by JelteF — APPROVED** [review]
I feel like it might be possible to move this `pselect`+(`un`)`block_signals` code to a shared function, since both places look so similar to the other one. If that's not the case for whatever reason feel free to merge this. It looks like a good change.

However, like discussed in chat though, this doesn't fix the issue described in https://github.com/citusdata/pg_auto_failover/issues/392. So another fix for that one is still needed.

**Comment by DimCitus** [comment]
> I feel like it might be possible to move this `pselect`+(`un`)`block_signals` code to a shared function, since both places look so similar to the other one. If that's not the case for whatever reason feel free to merge this. It looks like a good change.

I have now refactored the code in a way that allows better code share / re-use. This needs some more infrastructure, as in context structures and function pointers, to make it for the lack of closures and dynamic dispatch in C.

File: src/bin/pg_autoctl/monitor.c [hunk 1]
```diff
@@ -2982,25 +3031,28 @@ monitor_stop_maintenance(Monitor *monitor, int nodeId)
...
  * it's expected that the caller keeps polling the results to drain the queue
  * of notifications received from the previous calls loop.
  */
-bool
-monitor_get_notifications(Monitor *monitor, int timeoutMs)
+static bool
+monitor_process_notifications(Monitor *monitor,
+							  int timeoutMs,
+							  char *channels[],
+							  void *NotificationContext,
```
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:3050]: Looks like a type otherwise:
```suggestion
							  void *notificationContext,
```

**Review by JelteF — APPROVED** [review]

File: src/bin/pg_autoctl/monitor.c (hunk 1 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c]: Good point, done now.

**Pull request merged**